### PR TITLE
Add support for machine independent calculation for metrics data

### PR DESCRIPTION
**Pull request by joshuaranjan**
Resolves: #94

### CODE REVIEW

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (jandres742) [inline scripts/tools/EXT_Exp_MetricExportData.rst:16]: Data Experimental Extension

File: scripts/tools/metricExportData.yml
Comment (jandres742) [inline scripts/tools/metricExportData.yml:70]: missing quote

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (joshuaranjan) [reply scripts/tools/EXT_Exp_MetricExportData.rst]: updated

File: scripts/tools/metricExportData.yml
Comment (joshuaranjan) [reply scripts/tools/metricExportData.yml]: Fixed it.
Thanks

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (matcabral) [inline scripts/tools/EXT_Exp_MetricExportData.rst:45]: minor:  `Allocate buffer ..`

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (matcabral) [inline scripts/tools/EXT_Exp_MetricExportData.rst:56]: `machine independent` a bit ambiguous. How about saying something like ` .. how to perform metrics calculation of collected data, which can be done in a different system` ?

File: scripts/tools/metricExportData.yml
Comment (matcabral) [inline scripts/tools/metricExportData.yml:120]: IMHO would start by describing what the API does, then add the relations. That seems clearer. 

`Calculate metrics values with raw data returned by $GetExportDataExp, which can be done in different node that where data was collected. This function is similar to $tMetricGroupCalculateMultipleMetricValuesExp except it calculates from exported metric data.`

File: scripts/tools/metricExportData.yml
Comment (matcabral) [reply scripts/tools/metricExportData.yml:79]: ` don't need to have GPUs` 
we can't say GPUs, maybe accelerator.

File: scripts/tools/metricExportData.yml
Comment (joshuaranjan) [reply scripts/tools/metricExportData.yml:121]: Thanks. Please check the updated one.

File: scripts/tools/metricExportData.yml
Comment (joshuaranjan) [reply scripts/tools/metricExportData.yml]: Sure .. Modified for it

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (joshuaranjan) [reply scripts/tools/EXT_Exp_MetricExportData.rst]: Ok . Modified

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (joshuaranjan) [reply scripts/tools/EXT_Exp_MetricExportData.rst]: Updated. Thanks

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (joshuaranjan) [reply scripts/tools/EXT_Exp_MetricExportData.rst]: Updated

File: scripts/tools/metricExportData.yml
Comment (joshuaranjan) [reply scripts/tools/metricExportData.yml]: updated

**Review by matcabral — APPROVED** [review]

File: scripts/tools/common.yml
Comment (joshuaranjan) [reply scripts/tools/common.yml]: Fixed. Thanks

**Comment by jfedorov** [comment]
> This approach is similar to the existing API design of choosing a metric to collect on a platform. Since metric(name, description) are platform specific, the user calls a zetMetricGroupGet, reads its properties and does a string-compare using the metric name or its properties to choose their interested metric for collection.

Since derived metric calculation from raw metrics is also platform specific, they are represented as strings with description so that another vendor implementatation could choose to use a different data format for representing the calculation. So here we do not specify the format explicitly.
<
Yes. Agree on the metrics as char strings. 
The concern here is Math Operations (to be used in Equations) are not Platform specific. 
The Equation types: MetricCalculationEquation, ReadEquations are also to be specified. 
Let's make an API as less obscure as possible,

File: scripts/tools/metricExportData.yml
Comment (joshuaranjan) [reply scripts/tools/metricExportData.yml]: Thanks for the feedback.
To allow the data format to be strongly typed, an implementation specific header file would be used to describe it.
This would allow the exported data access to be easier to read and quicker to process.
Also this header file could support backward compatibility of data formats and version reporting for maintainability and error reporting

**Comment by joshuaranjan** [comment]
@jfedorov @matcabral 
Could you please review and approve if the L0 spec changes are OK.
Thanks!

File: scripts/tools/metricExportData.yml
Comment (jandres742) [inline scripts/tools/metricExportData.yml:109]: this API has two pCount parameters, so, should we add some of the same clarifications we have for https://github.com/oneapi-src/level-zero-spec/pull/96 ?

File: scripts/tools/metricExportData.yml
Comment (jandres742) [inline scripts/tools/metricExportData.yml:110]: these two pCount are not marked as optional, so they must be passed always, at the same time?

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (joshuaranjan) [reply scripts/tools/EXT_Exp_MetricExportData.rst:29]: We are passing a driver handle as the first parameter, so considered ze as the prefix
It would be better to have zet. 
But would it be OK to have zet prefix and not have MetricHandles passed ?

File: scripts/tools/metricExportData.yml
Comment (joshuaranjan) [reply scripts/tools/metricExportData.yml:110]: Yes.. 
We need to query for the 
* number of sets
* total number of metrics available

So that application can allocate accordingly

File: scripts/tools/metricExportData.yml
Comment (joshuaranjan) [reply scripts/tools/metricExportData.yml:109]: Would it be OK if we could add changes from #96 after it is merged, since the objective of #96 is different from this one ?

File: scripts/tools/metricExportData.yml
Comment (joshuaranjan) [reply scripts/tools/metricExportData.yml]: fixed. Thanks

**Comment by jfedorov** [comment]
thank you. I approve it on behalf of VTune team

File: scripts/tools/EXT_Exp_MetricExportData.rst
Comment (joshuaranjan) [reply scripts/tools/EXT_Exp_MetricExportData.rst:29]: Hi @jandres742 ,

>We are passing a driver handle as the first parameter, so considered ze as the prefix
It would be better to have zet.
But would it be OK to have zet prefix and not have MetricHandles passed ?

Could you please provide your opinion on the above clarification.

Thanks

Comment (jandres742) [reply scripts/tools/EXT_Exp_MetricExportData.rst:29]: zet is not about Metrics Handles but about whether the API belongs to Core or Tools. In this case, this API is expected to be used only for Tools, so zet would be better in my opinion.

Comment (joshuaranjan) [reply scripts/tools/EXT_Exp_MetricExportData.rst]: Thanks @jandres742 
Renamed the API according to the module it belongs to.

**Review by jandres742 — COMMENTED** [review]
+1

@wdamon-intel : please review.

**Comment by jandres742** [comment]
@matcabral : please review.

File: scripts/tools/metricExportData.yml
Comment (matcabral) [inline scripts/tools/metricExportData.yml:49]: minor: the description should be more explicit saying where are the reports skipped taken from. Beginning, end, middle?

**Comment by joshuaranjan** [comment]
Hi @wdamon-intel @jandres742 
Would it be possible to get this in 1.6.x ?

**Comment by wdamon-intel** [comment]
@joshuaranjan I've updated the project and milestone to v1.6.x.

**Review by jandres742 — APPROVED** [review]

**Pull request merged**
